lib/tom_queue/worker_supervisor.rb

```diff
@@ -0,0 +1,67 @@
+module TomQueue
+# class WorkerPool?
+  class WorkerSupervisor
+
+    attr_reader :processes
+
+    def initialize(processes = default_processes)
+      @processes = processes
+      setup_signal_handling
+    end
+
+    def default_processes
+      # TODO: deferred worker
+      {
+        "worker1" => -> { TomQueue::Worker.new.start },
+        "worker2" => -> { TomQueue::Worker.new.start },
+        "worker3" => -> { TomQueue::Worker.new.start },
+        "worker4" => -> { TomQueue::Worker.new.start },
+        "scheduler" => -> { TomQueue::Worker.new.start }
+      }
+    end
+
+    def setup_signal_handling
+      ["SIGINT","SIGTERM"].each do |signal|
+        Signal.trap(signal) do
+          raise SignalException, "SIGTERM"
+        end
+      end
+    end
+
+    def start_workers
+      pids = {}
+      # TODO: before_fork
+      loop do
+        processes_to_start = (processes.keys - pids.values)
+        processes_to_start.each do |name|
+          $stderr.puts "Starting '#{name}'..."
+          pid = fork_worker(processes[name])
+          pids[pid] = name
+        end
+
+        pid, status = wait_for_child_process_to_exit
+
+        if name = pids.delete(pid)
+          $stderr.puts "#{name} process terminated (#{pid}) with status #{status}"
+        end
+        sleep 1
+      end
+    end
+
+    private
+
+    def fork_worker(process)
```

Comment:
I'd be tempted to write a "child process" method then you can call (`fork(&method(:child_process))`) in the fork loop without the various nested blocks, etc